### PR TITLE
Add LanguageVersion for C# 12

### PR DESCRIPTION
**Pull request by jcouv**
Fixes https://github.com/dotnet/roslyn/issues/69109

I also pinged Lee by email to add C# 12 to [manifest](https://github.com/dotnet/core-private/pull/176/files)

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Can you update LanguageVersionExtensions.cs on the IDE side when you do this? We also need to update all the tests taht use LangVersion.Preview in order to test C#12 stuff on the IDE :)

**Comment by jcouv** [comment]
> Can you update LanguageVersionExtensions.cs on the IDE side when you do this?

I did already. Did I miss something?

**Comment by jcouv** [comment]
@CyrusNajmabadi I'd like to add a test that fails if someone forgot to update `LanguageVersionExtensions` when we ramp versions up in the compiler. Where would be a good place to put such a test?

**Comment by jcouv** [comment]
> We also need to update all the tests taht use LangVersion.Preview in order to test C#12 stuff on the IDE :)

I'm confused. Tests meant for C# 12 should have been using `LanguageVersionExtensions.CSharpNext`. Test that use `LangVersion.Preview` have a different meaning, they keep rolling forward forever, always staying on preview.

**Comment by CyrusNajmabadi** [comment]
Microsoft.CodeAnalysis.CSharp.Workspaces.UnitTests would probably be teh right place for that test.

**Comment by CyrusNajmabadi** [comment]
> I'm confused. Tests meant to use C# 12 specifically should have been using LanguageVersionExtensions.CSharpNext.

I don't think any of our tests do that.  I think tehy all just hardcode in LangVersion.Preview since htat's what CSharpNext was always set to.

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler @CyrusNajmabadi for review. Thanks

File: src/Analyzers/CSharp/Tests/UseCollectionExpression/UseCollectionExpressionForArrayTests.cs [hunk 1]
```diff
@@ -52,7 +52,7 @@ class C
                     int[] i = [1, 2, 3];
                 }
                 """,
-            LanguageVersion = LanguageVersionExtensions.CSharpNext,
+            LanguageVersion = LanguageVersion.CSharp12,
```
Comment (jaredpar) [inline src/Analyzers/CSharp/Tests/UseCollectionExpression/UseCollectionExpressionForArrayTests.cs:55]: Why do we pick `CSharp12` here vs. `LatestMajor`?

**Review by jaredpar — APPROVED** [review]

File: src/Analyzers/CSharp/Tests/UseCollectionExpression/UseCollectionExpressionForArrayTests.cs (hunk 1 above)
Comment (jcouv) [reply src/Analyzers/CSharp/Tests/UseCollectionExpression/UseCollectionExpressionForArrayTests.cs:55]: The change from CSharpNext to CSharp12 is designed to be mechanical, so not a lot of scrutiny was given to individual tests (I expect that would have been discussed when the test was written).
This specific test is call `TestInCSharp12`, so CSharp12 seems right.

File: src/EditorFeatures/CSharpTest/Formatting/CodeCleanupTests.cs [hunk 2]
```diff
@@ -442,7 +442,7 @@ internal class Program
         private void Method()
         {
             Console.WriteLine();
-            List<int> list = new();
+            List<int> list = [];
```
Comment (RikkiGibson) [inline src/EditorFeatures/CSharpTest/Formatting/CodeCleanupTests.cs:445]: oh wow, we have this feature already?

**Review by RikkiGibson — APPROVED** [review]

**Comment by jcouv** [comment]
@CyrusNajmabadi for review of IDE side. Thanks

**Comment by CyrusNajmabadi** [comment]
Looking.

File: src/Workspaces/CSharpTest/CSharpCommandLineParserServiceTests.cs [hunk 3]
```diff
@@ -4,7 +4,9 @@
 
 #nullable disable
 
+using System;
 using System.IO;
+using System.Linq;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharpTest/CSharpCommandLineParserServiceTests.cs:9]: que?

**Review by CyrusNajmabadi — APPROVED** [review]

**Comment by aetos382** [comment]
Why haven't the *Lambda default parameters* been moved to the C# 12.0 section?

**Comment by jcouv** [comment]
@aetos382 thanks for catching this oversight. I'll fix immediately

**Pull request merged**